compiler/qsc_eval/src/lib.rs

```diff
@@ -594,6 +561,52 @@ impl State {
                         }
                     }
                 }
+                Some(ExecGraphNode::Jump(idx)) => {
+                    self.idx = *idx;
+                    continue;
+                }
+                Some(ExecGraphNode::JumpIf(idx)) => {
+                    let cond = self.curr_val == Some(Value::Bool(true));
+                    if cond {
+                        self.idx = *idx;
+                    } else {
+                        self.idx += 1;
+                    }
+                    continue;
+                }
+                Some(ExecGraphNode::JumpIfNot(idx)) => {
+                    let cond = self.curr_val == Some(Value::Bool(true));
+                    if cond {
+                        self.idx += 1;
+                    } else {
+                        self.idx = *idx;
+                    }
+                    continue;
+                }
+                Some(ExecGraphNode::Store) => {
+                    self.push_val();
+                    self.idx += 1;
+                    continue;
+                }
+                Some(ExecGraphNode::Unit) => {
+                    self.idx += 1;
+                    self.set_curr_val(Value::unit());
+                    continue;
+                }
+                Some(ExecGraphNode::Ret) => {
+                    self.leave_frame();
+                    env.leave_scope();
+                    continue;
+                }
+                None => {
+                    // We have reached the end of the current graph without reaching an explicit return node,
+                    // usually indicating the partial execution of a single sub-expression.
+                    // This means we should pop the execution graph but not the current environment scope,
+                    // so bound variables are still accessible after completion.
+                    self.exec_graph_stack.pop();
```

Comment:
Should we add something like this to verify that we have effectively reached the end of the current graph?

Reply:
I don't think this would work, because the graph is read-only, so traversing it doesn't shorten it. When we pop it, it will still have all the contents it had before.